src/pytest_codspeed/plugin.py

```diff
@@ -172,86 +174,72 @@ def pytest_collection_modifyitems(
+        finally:
+            # Ensure instrumentation is stopped even if the test failed
+            lib.stop_instrumentation()
+            uri = get_git_relative_uri(nodeid, config.rootpath)
+            lib.dump_stats_at(uri.encode("ascii"))
+    finally:
+        # Ensure GC is re-enabled even if the test failed
+        if is_gc_enabled:
+            gc.enable()
+
+
+def wrap_runtest(
+    lib: LibType,
+    nodeid: str,
+    config: pytest.Config,
+    fn: Callable[P, T],
+) -> Callable[P, T]:
+    @functools.wraps(fn)
+    def wrapped(*args: P.args, **kwargs: P.kwargs) -> T:
+        return _run_with_instrumentation(lib, nodeid, config, fn, *args, **kwargs)
 
-    return result
+    return wrapped
 
 
 @pytest.hookimpl(tryfirst=True)
 def pytest_runtest_protocol(item: pytest.Item, nextitem: pytest.Item | None):
     plugin = get_plugin(item.config)
     if not plugin.is_codspeed_enabled or not should_benchmark_item(item):
-        return (
-            None  # Defer to the default test protocol since no benchmarking is needed
-        )
+        # Defer to the default test protocol since no benchmarking is needed
+        return None
 
     if has_benchmark_fixture(item):
-        return None  # Instrumentation is handled by the fixture
+        # Instrumentation is handled by the fixture
+        return None
 
     plugin.benchmark_count += 1
-    if not plugin.should_measure:
-        return None  # Benchmark counted but will be run in the default protocol
-
-    # Setup phase
-    reports = []
-    ihook = item.ihook
-    ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
-    setup_call = pytest.CallInfo.from_call(
-        lambda: ihook.pytest_runtest_setup(item=item, nextitem=nextitem), "setup"
-    )
-    setup_report = ihook.pytest_runtest_makereport(item=item, call=setup_call)
-    ihook.pytest_runtest_logreport(report=setup_report)
-    reports.append(setup_report)
-    # Run phase
-    if setup_report.passed and not item.config.getoption("setuponly"):
-        assert plugin.lib is not None
-        runtest_call = pytest.CallInfo.from_call(
-            lambda: _run_with_instrumentation(
-                plugin.lib, item.nodeid, item.config, item.runtest
-            ),
-            "call",
-        )
-        runtest_report = ihook.pytest_runtest_makereport(item=item, call=runtest_call)
-        ihook.pytest_runtest_logreport(report=runtest_report)
-        reports.append(runtest_report)
-
-    # Teardown phase
-    teardown_call = pytest.CallInfo.from_call(
-        lambda: ihook.pytest_runtest_teardown(item=item, nextitem=nextitem), "teardown"
-    )
-    teardown_report = ihook.pytest_runtest_makereport(item=item, call=teardown_call)
-    ihook.pytest_runtest_logreport(report=teardown_report)
-    reports.append(teardown_report)
-    ihook.pytest_runtest_logfinish(nodeid=item.nodeid, location=item.location)
+    if not plugin.should_measure or not plugin.lib:
```

Comment:
This `not plugin.lib` additional condition seem unnecessary here. It wasn't handled before so I'd prefer not changing that.
```suggestion
    if not plugin.should_measure:
```

Reply:
I saw it created a typing issue. I added this to fix the issue(this was there in the previous implementation as well):
```
    assert plugin.lib is not None
```